### PR TITLE
feat(artifacts): add optional timeout parameter to artifacts wait()

### PR DESCRIPTION
**Pull request by estellazx**
Fixes WB-8590

Description
-----------
Add optional timeout parameter to Artifact.wait()

Testing
-------
Unit tests testing that the wait() fails if the timeout runs out and wait() successes when there is no timeout or wait is finished faster than timeout

Checklist
-------
- Link to ticket: https://wandb.atlassian.net/jira/software/c/projects/WB/boards/62?modal=detail&selectedIssue=WB-8590

### CODE REVIEW

File: tests/unit_tests/test_wandb_run.py [hunk 1]
```diff
@@ -454,3 +455,29 @@ def test_settings_unexpected_args_telemetry(runner, relay_server, capsys, user):
...
+    elapsed_time = time.time() - start_timestamp
+    assert elapsed_time < timeout * (1 + leeway)
+    run.finish()
+
+
+@pytest.mark.parametrize("timeout", [0, 2])
+def test_artifact_wait_failure(user, test_settings, timeout):
+    # Test to expect TimeoutError when wait timeout is reached and large image
+    # wasn't uploaded yet
+    large_image = wandb.Image(np.random.randint(0, 255, (10000, 10000, 3)))
```
Comment (KyleGoyette) [inline tests/unit_tests/test_wandb_run.py:477]: Oh wow this is huge. Someone else will need to jump in here, will this test be okay to run on circle? @wandb/sdk-team

Comment (KyleGoyette) [reply tests/unit_tests/test_wandb_run.py:477]: Maybe testing with 0 is enough, even though its kind of a corner case

File: wandb/sdk/wandb_run.py [hunk 2]
```diff
@@ -3575,9 +3575,14 @@ def __getattr__(self, item: str) -> Any:
         self._assert_instance()
         return getattr(self._instance, item)
 
-    def wait(self) -> ArtifactInterface:
+    def wait(self, wait_timeout_secs: int = None) -> ArtifactInterface:
         if not self._instance:
-            resp = self._future.get().response.log_artifact_response
+            future_get = self._future.get(wait_timeout_secs)
+            if not future_get:
+                raise TimeoutError(
```
Comment (KyleGoyette) [inline wandb/sdk/wandb_run.py:3582]: @speezepearson is the desired outcome a raised error?

File: wandb/sdk/wandb_artifacts.py [hunk 3]
```diff
@@ -60,7 +60,7 @@
 
 # This makes the first sleep 1s, and then doubles it up to total times,
 # which makes for ~18 hours.
-_REQUEST_RETRY_STRATEGY = urllib3.util.retry.Retry(
+_REQUEST_RETRY_STRATEGY = requests.packages.urllib3.util.retry.Retry(
```
Comment (KyleGoyette) [inline wandb/sdk/wandb_artifacts.py:63]: Are we converting to requests from urllib3?

**Review by KyleGoyette — COMMENTED** [review]
Generally LGTM, concerned about just how big the array is for the test failure with the 2 second timeout. Maybe we should just test the 0 second time out?

File: wandb/sdk/wandb_artifacts.py (hunk 3 above)
Comment (estellazx) [reply wandb/sdk/wandb_artifacts.py:63]: I didn't actively make this change, this might have been caused by a merge conflict, I'll change it back to match master.

File: tests/unit_tests/test_wandb_run.py (hunk 1 above)
Comment (estellazx) [reply tests/unit_tests/test_wandb_run.py:477]: yeah I had trouble getting it to timeout so I ended up using that image size. Let me try making it smaller

File: wandb/sdk/wandb_run.py (hunk 2 above)
Comment (speezepearson) [reply wandb/sdk/wandb_run.py:3582]: I think that's the only way to make this change backwards-compatible! Traditionally (like, in the [`threading`](https://docs.python.org/3/library/threading.html) module), a timeout results in returning `False` instead of `True`; but we can't change this return type. (Or, well, I guess we could return an `Optional[ArtifactInterface]`, and only return None if it times out? I wouldn't object to that, but I don't see any advantages it has over the raise-an-exception solution.)

File: tests/unit_tests/test_wandb_run.py (hunk 1 above)
Comment (speezepearson) [reply tests/unit_tests/test_wandb_run.py:477]: Aha, a timeout of 0 means "no timeout" -- see [this condition](https://github.com/wandb/client/blob/4c87b3676c464414d31550e52564dd0c7fec577a/wandb/sdk/interface/message_future_poll.py#L46):
```python
            if timeout and start_time - now_time > timeout:
                done = True
                continue
```
`timeout=0` and `timeout=None` will both fail that check, because they're both falsey.

Maybe try a timeout of `1e-12`? Then I bet you won't need the image to be so huge.

File: wandb/sdk/wandb_run.py (hunk 2 above)
Comment (speezepearson) [inline wandb/sdk/wandb_run.py:3582]: Could we define our own `TimeoutError` type instead? From [the docs for the stdlib's TimeoutError](https://docs.python.org/3/library/exceptions.html#TimeoutError), it sounds like it's meant to represent low-level system errors, rather than issues in application logic.

**Review by speezepearson — CHANGES_REQUESTED** [review]

File: tests/unit_tests/test_wandb_run.py (hunk 1 above)
Comment (estellazx) [reply tests/unit_tests/test_wandb_run.py:477]: I believe the wait logic passed in from artifact.wait() is actually in https://github.com/wandb/wandb/blob/4c87b3676c464414d31550e52564dd0c7fec577a/wandb/sdk/interface/router.py#L35
So timeout=None still works

File: wandb/sdk/wandb_run.py (hunk 2 above)
Comment (KyleGoyette) [reply wandb/sdk/wandb_run.py:3582]: Hmm, do we have another error we raise when existing timeouts elapse? I would be surprised if we didn't already have this exact logic somewhere else.

Comment (KyleGoyette) [reply wandb/sdk/wandb_run.py:3582]: Either is fine by me!

File: tests/unit_tests/test_wandb_run.py [hunk 4]
```diff
@@ -454,3 +455,29 @@ def test_settings_unexpected_args_telemetry(runner, relay_server, capsys, user):
             # whose field 2 corresponds to unexpected arguments in Settings
             assert 2 in telemetry.get("11", [])
             run.finish()
+
+
+def test_artifact_wait_success(user, test_settings):
```
Comment (dmitryduev) [inline tests/unit_tests/test_wandb_run.py:460]: I think these tests should go to `tests/unit_tests/test_wandb_artifacts_full.py`

You probably want to use the `wandb_init` fixture here instead of these two (and perhaps also the `relay_server` server one to check on things, although some extra work might be needed there to resolve artifacts-related stuff, which we have to do anyways at some point), see here for a minimal example: https://github.com/wandb/wandb/blob/master/tests/unit_tests/test_wandb.py#L324

File: tests/unit_tests/test_wandb_run.py (hunk 1 above)
Comment (dmitryduev) [reply tests/unit_tests/test_wandb_run.py:477]: Yea, I agree with @KyleGoyette -- this might sporadically blow up the Docker container that Circle assigns to the job. How about mocking some artifact code internals with some sleep statements?

File: wandb/sdk/wandb_run.py (hunk 2 above)
Comment (estellazx) [reply wandb/sdk/wandb_run.py]: Is there a place to see all the wandb.errors? otherwise I just defined a **WaitTimeoutError**. Most timeout related errors either return ValueErrors or TimeoutErrors from what I've gathered after quickly searching around in this repo

File: tests/unit_tests/test_wandb_artifacts_full.py [hunk 5]
```diff
@@ -193,3 +194,29 @@ def make_table():
     artifact2.add(t1, "t2")
     assert artifact2.manifest.entries["t2.table.json"].ref is not None
     run.finish()
+
+
+def test_artifact_wait_success(wandb_init):
+    # Test artifact wait() timeout parameter
+    timeout = 2
```
Comment (speezepearson) [inline tests/unit_tests/test_wandb_artifacts_full.py:201]: IIUC this timeout is supposed to be "large enough that the artifact-upload will always complete in less than this many seconds"; but 2sec is short enough that it's within the range of plausible execution times. If there's no countervailing reason to keep it small, could we make it Obviously Incredibly Big, like 60, or 3600, or 9999999?

(I'm partly worried about hiccups in CI scheduling causing this test to flake, and partly just trying to make the code as obvious as possible)

File: tests/unit_tests/test_wandb_artifacts_full.py [hunk 6]
```diff
@@ -193,3 +194,29 @@ def make_table():
...
+    run = wandb_init()
+    artifact = wandb.Artifact("art", type="dataset")
+    start_timestamp = time.time()
+    run.log_artifact(artifact).wait(wait_timeout_secs=timeout)
+    elapsed_time = time.time() - start_timestamp
+    assert elapsed_time < timeout * (1 + leeway)
+    run.finish()
+
+
+@pytest.mark.parametrize("timeout", [0, 1e-1])
```
Comment (speezepearson) [inline tests/unit_tests/test_wandb_artifacts_full.py:212]: IIUC this timeout is supposed to be "small enough that the artifact-upload will never complete in less than this many seconds"; but 0.1sec is long enough that it's within the range of plausible execution times. If there's no countervailing reason to keep it large, could we make it Obviously Incredibly Small, like 1e-6, or 1e-12?

(I'm partly worried about hiccups in CI scheduling causing this test to flake, and partly just trying to make the code as obvious as possible)

File: wandb/sdk/wandb_run.py [hunk 7]
```diff
@@ -3586,9 +3592,14 @@ def __getattr__(self, item: str) -> Any:
         self._assert_instance()
         return getattr(self._instance, item)
 
-    def wait(self) -> ArtifactInterface:
+    def wait(self, wait_timeout_secs: int = None) -> ArtifactInterface:
```
Comment (speezepearson) [inline wandb/sdk/wandb_run.py:3595]: Nit: could we name this `timeout` or `timeout_sec`? The `wait` seems redundant, and we can't really change this after release.

File: wandb/sdk/wandb_artifacts.py [hunk 8]
```diff
@@ -647,9 +647,9 @@ def delete(self) -> None:
             "Cannot call delete on an artifact before it has been logged or in offline mode"
         )
 
-    def wait(self) -> ArtifactInterface:
+    def wait(self, wait_timeout_secs: int = None) -> ArtifactInterface:
```
Comment (speezepearson) [inline wandb/sdk/wandb_artifacts.py:650]: (as below re: argument naming)

**Review by speezepearson — CHANGES_REQUESTED** [review]

**Review by speezepearson — APPROVED** [review]
🚀

File: wandb/sdk/wandb_run.py [hunk 9]
```diff
@@ -138,6 +138,12 @@ class TeardownHook(NamedTuple):
     stage: TeardownStage
 
 
+class WaitTimeoutError(Exception):
```
Comment (kptkin) [inline wandb/sdk/wandb_run.py:141]: Do you think it would be good idea to move this to `errors/__init__.py`

Comment (estellazx) [reply wandb/sdk/wandb_run.py]: moved WaitTImeoutError to errors/__init__.py
